### ISSUE TITLE
Incorrect Unused Language Extension Warning when using quasiquotes

### ISSUE THREAD

**Issue by waddlaw** [ndmitchell/hlint#391]
```haskell
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE OverloadedStrings #-}

import Text.RawString.QQ

import Data.Text (Text)
import Data.Yaml (decode)

f :: Maybe Text
f = decode [r||]
```

HLint produces this incorrect output:

```haskell
./error.hs:2:1: Warning: Unused LANGUAGE pragma
Found:
  {-# LANGUAGE OverloadedStrings #-}
Why not remove it.
```

HLint version:

```bash
curl -sL https://raw.github.com/ndmitchell/hlint/master/misc/travis.sh | sh -s . -V
Downloading and running hlint...
######################################################################## 100.0%
HLint v2.0.9, (C) Neil Mitchell 2006-2017
```

local HLint is fine working:

```bash
$ hlint --version
HLint v1.9.35, (C) Neil Mitchell 2006-2016

$ hlint .
No hints
```

**Comment by ndmitchell** [comment]
Hmm, I thought the code inside a quasi quote or other template Haskell construct was mostly immune from which language features were in scope, but seemingly not. I wonder if this is a more general issue?

The local vs remote is merely because HLint on the server is a newer version which can detect otherwise unused `OverloadedStrings` extension pragmas.

**Comment by waddlaw** [comment]
reduced dependencies.

```haskell
-- A.hs
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE OverloadedStrings #-}

import Data.Text (Text)
import QQ (r)

f :: Text
f = [r||]
```

```haskell
-- Q.hs
module QQ (r) where

import Language.Haskell.TH
import Language.Haskell.TH.Quote

r :: QuasiQuoter
r = QuasiQuoter { quoteExp = return . LitE . StringL
                , quotePat = undefined
                , quoteType = undefined
                , quoteDec = undefined
                }
```

```bash
$ stack repl --resolver lts-9.5 Q.hs A.hs
*Main> :t f
f :: Text
*Main> f
""
```

I tried it with the latest version of local hlint.

```bash
$ hlint --version
HLint v2.0.9, (C) Neil Mitchell 2006-2017

$ hlint .
./A.hs:2:1: Warning: Unused LANGUAGE pragma
Found:
  {-# LANGUAGE OverloadedStrings #-}
Why not remove it.

1 hint
```

If delete the `OverloadedStrings` pragma according to the hint, it looks like this.

```bash
$ stack repl --resolver lts-9.5 Q.hs A.hs
A.hs:8:8: error:
    • Couldn't match expected type ‘Text’ with actual type ‘[Char]’
    • In the expression: ("")
      In an equation for ‘f’: f = ("")
Failed, modules loaded: QQ.
Loaded GHCi configuration from /tmp/ghci4802/ghci-script
```

**Comment by ndmitchell** [comment]
In general, if you use any QuasiQuotes or TemplateHaskell, then the generated code can make use of extensions will are required by the generated code, but there's no way to tell. That said, I think that if possible template haskell code should make use of as few extensions as possible - e.g. if `Text.RawString.QQ` is going to generate string literals, it should wrap them in `fromString` itself rather than turning on an extension to do so.

**Comment by waddlaw** [comment]
Ok, I understand.
Thank you. I will do it like that.
